articles/tutorial-qdk-grovers-search.md

```diff
@@ -11,49 +11,367 @@ title: Run Grover's search algorithm in Q# - Quantum Development Kit
+1. Start with a register of $n$ qubits initialized in the state $\ket{0}$ by applying $H$ to each qubit of the register.
+1. Prepare the register into a uniform superposition: $$|\psi\rangle=\frac{1}{N^{1 / 2}} \sum_{x=0}^{N-1}|x\rangle$$
+1. Apply $N_{\text{optimal}}$ times the following operations to the register:
+   1. The phase oracle $O_f$ that applies a conditional phase shift of $-1$ for the solution items.
+   1. Apply $H$ to each qubit of the register.
+   1. A conditional phase shift of $-1$ to every computational basis state except $\ket{0}$.
+   1. Apply $H$ to each qubit of the register.
+1. Measure the register to obtain the index of a item that's a solution with very high probability.
+1. Check if it's a valid solution. If not, start again.
 
-## Write the code
+## Write the code for Grover's algorithm
 
-1. Using the Quantum Development Kit, [create a new Q# project for the application](xref:microsoft.quantum.install-qdk.overview.standalone). Title the project `Grover`.
+Now let's see how to implement the code in Q#.
 
-1. Add the following code to the `Program.qs` file in your new project:
+### Grover's diffusion operator
+
+First, we are going to write an operation that applies the steps **b**, **c** and **d** of the loop. These steps
+are sometimes known as the Grover's diffusion operation.
+
+```qsharp
+operation ReflectAboutUniform(inputQubits : Qubit[]) : Unit {
+
+    within {
+        ApplyToEachA(H, inputQubits);
+        ApplyToEachA(X, inputQubits);
+    } apply {
+        Controlled Z(Most(inputQubits), Tail(inputQubits));
+    }
+
+}
+```
 
-    :::code language="qsharp" source="~/quantum/samples/algorithms/simple-grover/SimpleGrover.qs" range="4-":::
+In this operation we use the [within-apply](xref:microsoft.quantum.qsharp.conjugations) statement that implements the
+conjugation operations that occurs in the steps of the Grover's diffusion
+operation.
 
-1. To define the list that we're searching, create a new file `Reflections.qs`, and paste in the following code:
+> [!NOTE]
+> To learn more about conjugations in Q#, check the [conjugations
+> article in the Q# language guide](xref:microsoft.quantum.qsharp.conjugations).
 
-    :::code language="qsharp" source="~/quantum/samples/algorithms/simple-grover/Reflections.qs" range="4-":::
+You can check what each of the operations and functions used is by looking into
+the API documentation:
 
-    The `ReflectAboutMarked` operation defines the marked input that you are searching for: the string of alternating zeros and ones. This sample hard-codes the marked input, and can be extended to search for different inputs or generalized for any input.
+- [`ApplyToEachA`](xref:Microsoft.Quantum.Canon.ApplyToEachA)
+- [`Most`](xref:microsoft.quantum.arrays.most)
+- [`Tail`](xref:microsoft.quantum.arrays.tail)
 
-1. Next, run your new Q# program to find the item marked by `ReflectAboutMarked`.
+A good exercise to understand the code and the operations is to check with pen
```

Comment:
This could be a reasonable place to include screenshots  of running in an IQ# Notebook and using `DumpMachine` to view the internal state of the simulator before and after reflecting.